sklearn/decomposition/dict_learning.py

```diff
@@ -333,90 +334,6 @@ def sparse_encode(X, dictionary, gram=None, cov=None, algorithm='lasso_lars',
     return code
 
 
-def _update_dict(dictionary, Y, code, verbose=False, return_r2=False,
-                 random_state=None, positive=False):
-    """Update the dense dictionary factor in place.
-
-    Parameters
-    ----------
-    dictionary : array of shape (n_features, n_components)
-        Value of the dictionary at the previous iteration.
-
-    Y : array of shape (n_features, n_samples)
-        Data matrix.
-
-    code : array of shape (n_components, n_samples)
-        Sparse coding of the data against which to optimize the dictionary.
-
-    verbose:
-        Degree of output the procedure will print.
-
-    return_r2 : bool
-        Whether to compute and return the residual sum of squares corresponding
-        to the computed solution.
-
-    random_state : int, RandomState instance or None, optional (default=None)
-        If int, random_state is the seed used by the random number generator;
-        If RandomState instance, random_state is the random number generator;
-        If None, the random number generator is the RandomState instance used
-        by `np.random`.
-
-    positive : boolean, optional
-        Whether to enforce positivity when finding the dictionary.
-
-        .. versionadded:: 0.20
-
-    Returns
-    -------
-    dictionary : array of shape (n_features, n_components)
-        Updated dictionary.
-
-    """
-    n_components = len(code)
-    n_features = Y.shape[0]
-    random_state = check_random_state(random_state)
-    # Get BLAS functions
-    gemm, = linalg.get_blas_funcs(('gemm',), (dictionary, code, Y))
-    ger, = linalg.get_blas_funcs(('ger',), (dictionary, code))
-    nrm2, = linalg.get_blas_funcs(('nrm2',), (dictionary,))
-    # Residuals, computed with BLAS for speed and efficiency
-    # R <- -1.0 * U * V^T + 1.0 * Y
-    # Outputs R as Fortran array for efficiency
-    R = gemm(-1.0, dictionary, code, 1.0, Y)
-    for k in range(n_components):
-        # R <- 1.0 * U_k * V_k^T + R
-        R = ger(1.0, dictionary[:, k], code[k, :], a=R, overwrite_a=True)
-        dictionary[:, k] = np.dot(R, code[k, :])
-        if positive:
-            np.clip(dictionary[:, k], 0, None, out=dictionary[:, k])
-        # Scale k'th atom
-        # (U_k * U_k) ** 0.5
-        atom_norm = nrm2(dictionary[:, k])
-        if atom_norm < 1e-10:
-            if verbose == 1:
-                sys.stdout.write("+")
-                sys.stdout.flush()
-            elif verbose:
-                print("Adding new random atom")
-            dictionary[:, k] = random_state.randn(n_features)
```

Comment:
You're no longer doing normally distributed sampling, are you?

Reply:
You're right. We need some sort of transform. Any that you would recommend?

Reply:
Was looking at that previously. Seems they don't really expose that. Raised issue ( https://github.com/numpy/numpy/issues/11764 ) on this point. Interestingly this has [come up before]( https://groups.google.com/forum/#!topic/cython-users/9UGMi_b3tVo ). Guess that's how the uniform random distribution code came to be in `cd_fast.pyx`. Perhaps we could vendor randomkit for now?

As a "simple" thought, we could use the inverse of the normal distribution's CDF to map uniformly distributed numbers to normally distributed ones. This would require some implementation of `erfi` (or better yet `ndtri`). There is a Cython interface to these in SciPy, but it requires 0.19.0 at least. Perhaps we could vendor the relevant code (from Netlib Cephes)?

Reply:
Have included an implementation of the [Marsaglia polar method]( https://en.wikipedia.org/wiki/Marsaglia_polar_method ) (based off the [Box–Muller transform]( https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform )), which appears to be the same thing that NumPy's vendored randomkit is doing. Interestingly the implementation included here appears to be ~2x faster than NumPy's.

Reply:
Speed improvements on numpy's random library aren't that interesting: their version compatibility assurances force them to keep sub-optimal versions

Reply:
if you guys don't need this numerical binary compatibility with standard Numpy's random module, do you want to try our `mkl_random` instead? It's available on all the channels and you can `try: import mkl_random as rnd` to make it optionally drop-in replacement for numpy.random. 
Performance: https://conference.scipy.org/proceedings/scipy2017/pdfs/oleksandr_pavlyk.pdf @oleksandr-pavlyk

Reply:
Is there a Cython or C/C++ header that could be used? The main goal in this PR is to move as much as possible of this function into non-GIL holding Cython code to maximize performance.